src/core/meshcop/commissioner.cpp

```diff
@@ -81,6 +84,29 @@ Commissioner::Commissioner(Instance &aInstance)
     mCommissionerAloc.mScopeOverrideValid = true;
 }
 
+void Commissioner::SetState(otCommissionerState aState)
+{
+    VerifyOrExit(mState != aState);
+
+    mState = aState;
+
+    if (mStateCallback)
+    {
+        mStateCallback(mState, mCallbackContext);
```

Comment:
Do we have a specific use-case for the direct callback `mStateCallback`? 
Can we instead expect the user to use the `OT_CHANGED_COMMISSIONER_STATE` and `otCommissionerGetState()` ? The benefit of this is that we don't need to worry about re-entry to the code for the `OT_CHANGED...` model (since the callback is invoked from its own tasklet).

Reply:
The callback is really used for UI purposes.

Couple reasons to avoid `OT_CHANGED_COMMISSIONER_STATE`:

1. Thread devices generally will not implement the commissioner role. It would be nice not to have to reserve a state changed flag for a role that most devices will not implement.

2. The state changed mechanism may not capture all state transitions, especially when reentrancy occurs. This can make UI more difficult to implement or understand.

Reply:
Got it. Sounds good.